### PR TITLE
fix: offline storage anr issue and excessive string concatenation

### PR DESCRIPTION
**Pull request by ndesai-newrelic**
#139 #191

### CODE REVIEW

File: agent/src/main/java/com/newrelic/agent/android/NewRelic.java [hunk 1]
```diff
@@ -472,9 +472,11 @@ public static void recordMetric(String name, String category, int count, double
         StatsEngine.notice().inc(MetricNames.SUPPORTABILITY_API
                 .replace(MetricNames.TAG_NAME, "recordMetric"));
 
-        log.debug("NewRelic.recordMetric invoked for name " + name + ", category: " + category +
-                ", count: " + count + ", totalValue " + totalValue + ", exclusiveValue: " + exclusiveValue +
-                ", countUnit: " + countUnit + ", valueUnit: " + valueUnit);
+        StringBuilder logString = new StringBuilder();
```
Comment (cthomas-newrelic) [inline agent/src/main/java/com/newrelic/agent/android/NewRelic.java:475]: How does this reduce excessive string concatenation? IIRC, the complaint was the agent doing the work to build this excessive `DEBUG` level log message, even when logLevel was set to something below DEBUG. The change, imo, is to wrap the call in a conditional that checks the current log level first, before constructing the message.

Comment (ndesai-newrelic) [reply agent/src/main/java/com/newrelic/agent/android/NewRelic.java:475]: String builder is more efficient when two or more strings are concat, i agree need to add log level before concatenation happen.

File: agent/src/main/java/com/newrelic/agent/android/stores/SharedPrefsEventStore.java [hunk 2]
```diff
@@ -74,7 +74,7 @@ public void delete(AnalyticsEvent event) {
         try {
             synchronized (this) {
                 final SharedPreferences.Editor editor = sharedPrefs.edit();
-                editor.remove(event.getEventUUID()).commit();
+                editor.remove(event.getEventUUID()).apply();
```
[the remark this answers is not in the document]
Comment (ndesai-newrelic) [reply agent/src/main/java/com/newrelic/agent/android/stores/SharedPrefsEventStore.java:77]: this is created ANR if we are adding two many events on main thread.

File: agent/src/main/java/com/newrelic/agent/android/NewRelic.java [hunk 3]
```diff
@@ -472,9 +472,12 @@ public static void recordMetric(String name, String category, int count, double
         StatsEngine.notice().inc(MetricNames.SUPPORTABILITY_API
                 .replace(MetricNames.TAG_NAME, "recordMetric"));
 
-        log.debug("NewRelic.recordMetric invoked for name " + name + ", category: " + category +
-                ", count: " + count + ", totalValue " + totalValue + ", exclusiveValue: " + exclusiveValue +
-                ", countUnit: " + countUnit + ", valueUnit: " + valueUnit);
+        if(log.getLevel() >= AgentLog.AUDIT) {
```
Comment (cthomas-newrelic) [inline agent/src/main/java/com/newrelic/agent/android/NewRelic.java:475]: You replaced a `debug()` call with `AUDIT`. Is that intended?

**Review by cthomas-newrelic — APPROVED** [review]
🐑

File: agent/src/main/java/com/newrelic/agent/android/NewRelic.java (hunk 3 above)
Comment (ndesai-newrelic) [reply agent/src/main/java/com/newrelic/agent/android/NewRelic.java:475]: yes, you mentioned it here. https://github.com/newrelic/newrelic-android-agent/issues/139#issuecomment-1753256540

**Pull request merged**